src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/TraversingBlockBalancedTreeReader.java

```diff
@@ -0,0 +1,314 @@
+
+        // Second +1 because MathUtil.log computes floor of the logarithm; e.g.
+        // with 5 leaves you need a depth=4 tree:
+        return MathUtil.log(numLeaves, 2) + 2;
+    }
+
+    interface TraversalCallback
+    {
+        void onLeaf(int leafNodeID, long leafBlockFP, IntArrayList pathToRoot);
+    }
+
+    final class PackedIndexTree
+    {
+        private final byte[][] splitPackedValueStack;
+
+        private int nodeID;
+        // level is 1-based so that we can do level-1 w/o checking each time:
+        private int level;
+        private boolean leafNode;
+
+        // used to read the packed byte[]
+        private final ByteArrayDataInput in;
+        // holds the minimum (left most) leaf block file pointer for each level we've recursed to:
+        private final long[] leafBlockFPStack;
+        // holds the address, in the packed byte[] index, of the left-node of each level:
+        private final int[] leftNodePositions;
+        // holds the address, in the packed byte[] index, of the right-node of each level:
+        private final int[] rightNodePositions;
+        // true if the per-dim delta we read for the node at this level is a negative offset vs. the last split on this dim; this is a packed
+        // 2D array, i.e. to access array[level][dim] you read from negativeDeltas[level*numDims+dim].  this will be true if the last time we
+        // split on this dimension, we next pushed to the left subtree:
+        private final boolean[] negativeDeltas;
+        // holds the packed per-level split values; the run method uses this to save the cell min/max as it recurses:
+        private final byte[][] splitValuesStack;
+        // scratch value to return from getPackedValue:
+        private final BytesRef scratch;
+
+        PackedIndexTree()
+        {
+            int treeDepth = getTreeDepth();
+            splitPackedValueStack = new byte[treeDepth + 1][];
+            nodeID = 1;
+            level = 1;
+            splitPackedValueStack[level] = new byte[packedBytesLength];
+            leafBlockFPStack = new long[treeDepth + 1];
+            leftNodePositions = new int[treeDepth + 1];
+            rightNodePositions = new int[treeDepth + 1];
+            splitValuesStack = new byte[treeDepth + 1][];
+            negativeDeltas = new boolean[treeDepth + 1];
+
+            in = new ByteArrayDataInput(packedIndex);
+            splitValuesStack[0] = new byte[packedBytesLength];
+            readNodeData(false);
+            scratch = new BytesRef();
+            scratch.length = bytesPerValue;
+        }
+
+        public void pushLeft()
+        {
+            int nodePosition = leftNodePositions[level];
+            nodeID *= 2;
+            level++;
+            if (splitPackedValueStack[level] == null)
+                splitPackedValueStack[level] = new byte[packedBytesLength];
+            System.arraycopy(negativeDeltas, level - 1, negativeDeltas, level, 1);
+            assert !leafNode;
```

Comment:
nit: Here and in `pushRight()`, decent `assert` messages?

Reply:
Agreed. Firstly, I have removed the `leafNode` flag in favour of just using `isLeafNode`. I've also moved the assertion to the start of the method. The only reason I can find for the existence of the `leafNode` flag was that the `nodeID` changed above this assertion.